### PR TITLE
src/mon: "GLOBAL %RAW USED" of "ceph df" is not consistent with check…

### PR DESCRIPTION
**Pull request by liupan1111**
…_full_status

Fixes: http://tracker.ceph.com/issues/19487
Signed-off-by: Pan Liu <liupan1111@gmail.com>

### CODE REVIEW

**Comment by liupan1111** [comment]
The core reason is, In OSDService::check_full_status, it has comment: 
`We base ratio on kb_avail rather than kb_used because they can differ significantly ...`

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -2057,9 +2057,9 @@ void PGMap::dump_fs_stats(
         << stringify(si_t(osd_sum.kb_used*1024));
     float used = 0.0;
     if (osd_sum.kb > 0) {
-      used = ((float)osd_sum.kb_used / osd_sum.kb);
+      used = ((float)(osd_sum.kb - osd_sum.kb_avail) / osd_sum.kb);
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: note for posterity:

the following warning mentioned in http://tracker.ceph.com/issues/19487 comes from a jewel OSD:

> 2017-04-04 20:57:40:894659 [ERR] OSD full dropping all updates 51% full

see https://github.com/ceph/ceph/commit/80829d7805b1113d322509345372bf9c7984559b#diff-fa6c2eba8356ae1442d1bf749beacfdfR621

the way how we calculate the use ratio here is different from `OSDService::check_full_status()`. that's why we have the discrepancy here.

Comment (tchaikov) [inline src/mon/PGMap.cc]: @sage both changes (here and the one OSD) are from you, maybe you could help double check it?

**Comment by dzafman** [comment]
retest this please

**Comment by dzafman** [comment]
@liupan1111 See my comment in http://tracker.ceph.com/issues/19487.  The RAW USED% describes the value it provides.  The only time you could compare this percentage with the individual OSD percentage would be if all OSDs had the same usage and so the global usage would be expected to be the same.  This is rare and I'm hesitant to change the "RAW USED %" to be something other than what it claims to be.

**Comment by liupan1111** [comment]
@dzafman thanks, I agree that RAW USED% of global usage won't exactly match individual OSD percentage. Indeed, I just take "ceph df" as an example, in my case, I create a cluster with only one OSD. The more important is: "ceph osd df", I think "USED%" of each OSD should be consistent with check_full_status, but it is not now. I could create a new PR for "ceph osd df" if you agree.

**Comment by dzafman** [comment]
@liupan1111 I'm not sure what you are suggesting.   It is critical that check_full_status() takes into account only available space because sometimes avail is less than (total - used).  So I think we can't change check_full_status() and we don't want to change "RAW USED%"

**Comment by liupan1111** [comment]
@dzafman I don't mean to change check_full_status, or "RAW USED%", I mean to change the "USED%" in "ceph osd df"(not ceph df)

**Comment by liupan1111** [comment]
@tchaikov， what is your opinion?

**Comment by dzafman** [comment]
@liupan1111 @tchaikov There is no change required at all.   It turns out that _used_ is computed from _available_ so the percentages will be equivalent.  If they aren't that needs to be investigated.  Could there is a difference that needs to be fixed in rounding?

The check_full_status(), individual OSD and global (sum of all OSDs) stat information all originates from ObjectStore::statfs() calls and propagated in the OSDMap.

OSDService::update_osd_stat(...
...
  int r = osd->store->statfs(&stbuf);
...
  uint64_t bytes = stbuf.total;
  uint64_t used = bytes - stbuf.available;
  uint64_t avail = stbuf.available;

The comment in OSDService::check_full_status() about btrfs and used vs available is incorrect.  I'll remove it and change the computation in my out of space pull request: ceph/ceph#13968 (05b09fd95b56a070fa5c0f6d880c72e2dccf2ba3).

**Comment by tchaikov** [comment]
@liupan1111 can we keep this PR closed for now, and revisit this change if there is still discrepancies after @dzafman 's change gets merged?

@dzafman thanks for clear this confusion!

**Pull request closed without merging**